src/script/interpreter.cpp

```diff
@@ -1087,9 +1116,23 @@ class CTransactionSignatureSerializer {
  
             if (txTo.IsScVersion() )
             {
+                // Serialize CSW inputs
+                // In case of SIGHASH_ANYONECANPAY:
+                // * if Tx has Sc support version and nIn belongs to the CSW inputs - only the CSW input being signed is serialized
+                // * otherwise skip CSW inputs
+                // Otherwise - serialize all CSW inputs
+                unsigned int nCswInputs = fAnyoneCanPay ?
+                            (nIn < txTo.GetVin().size() ? 0 : 1) :
+                            txTo.GetVcswCcIn().size();
+                ::WriteCompactSize(s, nCswInputs);
+                for (unsigned int nCswInput = 0; nCswInput < nCswInputs; nCswInput++)
+                    SerializeCswInput(s, nCswInput, txTo, nType, nVersion);
+
                 // Serialize vccouts
                 unsigned int nCcOutputs = 0;
  
+                // Q: In case of fHashSingle we also serialize all CC outputs. Is it correct?
+                // Seems or we need to ingore all CC outputs, or consider them in a unique joined index of(vouts + cc_outs)
```

Comment:
I would consider CCOuts as a unique index of `vouts + cc_outs`. The reference for my (currently limited) understanding of signatures is [this](https://raghavsood.com/blog/2018/06/10/bitcoin-signature-types-sighash). In here it is stated that `SIGHASH_SINGLE` should cover all inputs and just one output, expressing the intention "I want to spend my coins only if some other parties are making a transfer too". This interpretation extends naturally to ccOutput as well and suggest the unique index solution

Reply:
to be discussed.

Reply:
`fHashSingle` usage with CC data must be explored deeply.
In current implementation, when CC outputs are **anyways** serialized, the corner case is the following:
- Tx has fixed set of inputs. Tx has not been signed yet.
- The owner of the first input define the first output and sign the input with `fHashSingle`, so don't care how the change is going to be spent.
- Then Tx was appended with the FT output. It invalidates the signature of the first input, because the message to sign was changed.

Another approach, is when CC outputs are **never** serialized for `fHashNone` and `fHashSingle` flags. The limitation is that we are not able to link the input signature to the CC output, but only to the single output. So no sidechains support for the `fHashSingle`.

We may think about the new signing flags, like "hash single for crosschain output`. But it has it's own problems and implementation complexity.

Reply:
REACHED THIS POINT